### PR TITLE
rocks: unable to install with tarantoolctl rocks on Linux

### LINKED ISSUE #39: Unable to install with tarantoolctl rocks on Linux
```
$ tarantoolctl rocks install mqtt
<..cut..>
Install the project...
-- Install configuration: "RelWithDebInfo"
-- Installing: /home/alex/tmp/.rocks/share/tarantool/rocks/mqtt/scm-1/lib/mqtt/driver.so
-- Set runtime path of "/home/alex/tmp/.rocks/share/tarantool/rocks/mqtt/scm-1/lib/mqtt/driver.so" to ""
-- Installing: /home/alex/tmp/.rocks/share/tarantool/rocks/mqtt/scm-1/lua/mqtt/init.lua
CMake Error at third_party/mosquitto/build/cmake_install.cmake:41 (file):
  file cannot create directory: /usr/local/etc/mosquitto.  Maybe need
  administrative privileges.
Call Stack (most recent call first):
  mqtt/cmake_install.cmake:85 (include)
  cmake_install.cmake:42 (include)


gmake: *** [Makefile:118: install] Error 1

Error: Build error: Failed installing.
```

### PR DESCRIPTION
**Pull request by vr009**
The problem from #39 is solved by adding EXCLUDE_FROM_ALL option.

On any platforms many targets from subdirectories are included into make command by default. They are not necessary to build driver.{so|dylib}. EXCLUDE_FROM_ALL option disables all targets at the beginning and only targets which are explicitly added to mqtt dependencies will be built.

Additionally static build for rocks set by default.

Fixes #39

### CODE REVIEW

**Review by ligurio — COMMENTED** [review]
Thanks for the patch!
Please limit a width of commit message by 72 symbols, see [1].

1. https://www.tarantool.io/en/doc/latest/dev_guide/developer_guidelines/#how-to-write-a-commit-message

**Comment by ligurio** [comment]
> The problem from #39 is solved by adding EXCLUDE_FROM_ALL option.

I see that issue #39 collects a number of problems:
1. installation of `/usr/local/etc/mosquitto` requires administrative privileges
2. error on linking executable `mosquitto_rr`

Seems your commit solves only the first problem. What about second one?

**Comment by ligurio** [comment]
> Additionally static build for rocks set by default.

What for?

**Comment by vr009** [comment]
> I see that issue #39 collects a number of problems:
> 
> 1. installation of `/usr/local/etc/mosquitto` requires administrative privileges
> 2. error on linking executable `mosquitto_rr`
> 
> Seems your commit solves only the first problem. What about second one?

1. The requirement of administrative privileges was caused by some internal targets from libmosquitto subdirectories. 
They are not necessary to build module's driver. `EXCLUDE_FROM_ALL` option disables all the targets in current subdirectory by default. So those targets, which are explicitly added to mqtt dependencies, will be built. 
2. Error on linking executable `mosquitto_rr` was solved here: eclipse/mosquitto#1453. 
We should update mosquitto to at least 1.6.8.

**Comment by vr009** [comment]
> > Additionally static build for rocks set by default.
> 
> What for?

The idea was, that installation from rocks is working everywhere without need to install any library manually. That is default behavior. But it is still possible to build the module with additional options for CMake, if it is needed.

**Comment by ligurio** [comment]
> > > Additionally static build for rocks set by default.
> > 
> > What for?
>
> The idea was, that installation from rocks is working everywhere without need to install any library manually. That is default  behavior. But it is still possible to build the module with additional options for CMake, if it is needed.

Let's add explanation to a commit message.

**Comment by vr009** [comment]
> > > > Additionally static build for rocks set by default.
> > > 
> > > 
> > > What for?
> > 
> > 
> > The idea was, that installation from rocks is working everywhere without need to install any library manually. That is default  behavior. But it is still possible to build the module with additional options for CMake, if it is needed.
> 
> Let's add explanation to a commit message.

Done.

File: mqtt/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@
 option(DOCUMENTATION "Build documentation?" OFF)
 
 option(WITH_STATIC_LIBRARIES "Build static versions of the libmosquitto/pp libraries?" OFF)
+option(WITH_PIC "Build the static library with PIC (Position Independent Code) enabled archives?" OFF)
```
Comment (ligurio) [inline mqtt/CMakeLists.txt:7]: What for do you add this option? There is nothing about `WITH_PIC` in commit message.

**Review by ligurio — CHANGES_REQUESTED** [review]

File: mqtt/CMakeLists.txt (hunk 1 above)
Comment (vr009) [reply mqtt/CMakeLists.txt:7]: This is related to [CMP0077](https://cmake.org/cmake/help/latest/policy/CMP0077.html). I've added description of this in the commit message.

Comment (Totktonada) [reply mqtt/CMakeLists.txt:7]: If we'll not define it as an option explicitly, the effect of `set(WITH_PIC ON)` (several lines below, under `if (STATIC_BUILD)`) will be discarded by `option(WITH_PIC <...> OFF)` in `third_party/mosquitto/lib/CMakeLists.txt`.

Comment (Totktonada) [reply mqtt/CMakeLists.txt:7]: As for me, the reason in the commit message is wrong. Without this change `tarantoolctl rocks make` does not work for me:

```
[100%] Linking C shared library driver.so
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: ../third_party/mosquitto/build/lib/libmosquitto.a(net_mosq.c.o): warning: relocation against `_ui_method' in read-only section `.text'
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: ../third_party/mosquitto/build/lib/libmosquitto.a(net_mosq.c.o): relocation R_X86_64_PC32 against symbol `tls_ex_index_mosq' can not be used when making a shared object; recompile with -fPIC
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
```

Now it is described like 'disable -fPIC by default', but:

1. The commit is about building from the rockspec and in context of `tarantoolctl rocks make` it is enabled. That's confusing.
2. It is already disabled by default when building with cmake directly.

Comment (vr009) [reply mqtt/CMakeLists.txt:7]: Fixed.

Comment (Totktonada) [reply mqtt/CMakeLists.txt:7]: > That is needed for CMake could not set this option off in CMake file of third
> party library while building the module statically when this option is enabled.

I doubt I read it correctly. Can you rephrase it in the form like the following:

> Without this change the behaviour is the following. If \<precondition\>, then \<bad result\>.

So it'll be easy to spot why it is needed.

Comment (vr009) [reply mqtt/CMakeLists.txt:7]: Fixed, now it look better.

Comment (Totktonada) [reply mqtt/CMakeLists.txt:7]: Looks okay for me now.

**Review by Totktonada — APPROVED** [review]
LGTM.

**Review by ligurio — APPROVED** [review]
LGTM

File: mqtt/CMakeLists.txt (hunk 1 above)
Comment (Totktonada) [reply mqtt/CMakeLists.txt:7]: (Will resolve this discussion.)
